### ISSUE TITLE
Tripal 4 You can add a field to a content type that does not support it

### ISSUE THREAD

**Issue by dsenalik** [tripal/tripal#1440]
# BUG/ERROR report

### Issue description
We will need a mechanism to prevent adding a field to a content type that does not support it.
I reported in these comments https://github.com/tripal/tripal/issues/1432#issuecomment-1454840026 and https://github.com/tripal/tripal/issues/1432#issuecomment-1455117899 and I posted issue #1436, but it is not a problem with the yaml file but with the field not validating that the field can be added to all base tables.

So briefly, we need to gracefully display an error and clean up the broken field if the base table selected when adding a field does not support the field.

#### Steps to reproduce
My example was adding the ```obi__organism``` field to ```project``` and getting this error:

```Warning: Undefined array key "foreign keys" in Drupal\tripal_chado\Plugin\Field\FieldType\obi__organism::tripalTypes() (line 89 of /var/www/drupal9/web/modules/contrib/tripal/tripal_chado/src/Plugin/Field/FieldType/obi__organism.php) ```

For ```obi__organism``` we could add something like the following, but is this correct?
```
    if (isset($base_schema_def['foreign keys']['organism']['columns'])) {
      $base_fk_col = array_keys($base_schema_def['foreign keys']['organism']['columns'])[0];
    }
    else {
      throw new \Exception(t('base table @base_table does not support adding this field',
        ['@base_table' => $base_table]));
    }
```
![proposal](https://user-images.githubusercontent.com/8419404/223208819-07050ddb-9e2e-47cc-b042-5f37c5fd4548.png)

**Comment by spficklin** [comment]
Working on this on branch `tv4g1-issue1440-limit_fields`

**Comment by spficklin** [comment]
I think I've found a way to fix this issue.  Configuration entities allow any module to add [3rd party settings](https://api.drupal.org/api/drupal/core!lib!Drupal!Core!Config!Entity!ThirdPartySettingsInterface.php/interface/ThirdPartySettingsInterface/8.2.x).  Since our content types extend the base classes of configuration entities that means we can add 3rd party settings to any content type.   So my suggestion is that we use this third party settings interface to add a setting that only the Tripal Chado module will use that basically maps a chado base table to each content type.  For example, The YAML for defining a content type would look like this:

```yaml
id: 'general_chado'
label: 'General Content Types (Chado)'
description: 'More generalized content types based on Chado, which all Tripal sites should have.'
content_types:

    -   label: Organism
        term: OBI:0100026
        help_text: Use the organism page for an individual living system, such as animal, plant, bacteria or virus,
        category: General
        id: organism
        title_format: "[organism_genus] [organism_species] [organism_infraspecific_type] [organism_infraspecific_name]"
        url_format: "organism/[TripalEntity__entity_id]"
        synonyms:
            - bio_data_1
        settings:
            chado_base_table: organism
```
Note the new `settings` section.   This will allow a field to know if it is compatible with an entity type because it can check if it supports the base table set in the content type.  The benefit to using a 3rd party setting is that if we add another backend storage (e.g. NoSQL or graph database) then they can set their own settings and ignore the chado base table setting.  

Aside from using the YAML when creating content types, programmatically adding a setting to an existing content type is as easy as this:

```php
  $entity_type_manager = \Drupal::entityTypeManager();
  $entity_type = $entity_type_manager->getStorage('tripal_entity_type')->load($bundle);
  $entity_type->setThirdPartySetting('tripal', $key, $value);
  $entity_type->save();
```

With this approach, the base table is set in the 3rd party settings of the content type, and therefore, we no longer need the select box for the base table when adding a field to a content type.  We know what it should be because of the setting.

However the base table setting in the content type doesn't fully fix the problem, we still need a way to prevent the user from adding a field to a content type that isn't compatible.  I propose we add a new function `ChadoFieldItemBase::isCompatible();`.   Each field will then be responsible for responding TRUE or FALSE if it compatible with an entity type.   This takes the burden away from the `ChadoFieldItemBase::storageSettingsForm()` form to try and figure it out for every possible field and passes the responsiblity to the field itself.  

Here's an example for the `chado_organism_type_default` field:

```php
  /**
   * {@inheritDoc}
   * @see \Drupal\tripal_chado\TripalField\ChadoFieldItemBase::isCompatible()
   */
  public function isCompatible(TripalEntityType $entity_type) : bool {

    // Get the base table for the content type.
    $base_table = $entity_type->getThirdPartySetting('tripal', 'chado_base_table');
    $chado = \Drupal::service('tripal_chado.database');
    $schema = $chado->schema();
    $base_table_def = $schema->getTableDef($base_table, ['format' => 'Drupal']);

    // If the base table has a foreign key to the organism table then this
    // content type is compatible.
    if (in_array('organism', $base_table_def['foreign keys'])) {
      return TRUE;
    }
    return FALSE;
  }
```
For the project linker the code could look like this:

```php
  /**
   * {@inheritDoc}
   * @see \Drupal\tripal_chado\TripalField\ChadoFieldItemBase::isCompatible()
   */
  public function isCompatible(TripalEntityType $entity_type) : bool {

    // Get the base table for the content type.
    $base_table = $entity_type->getThirdPartySetting('tripal', 'chado_base_table');

    // Is there a corresponding linker table?
    $object_table = self::$object_table;
    $linker = 'project_' . $object_table;

    /** @var \Drupal\tripal_chado\Database\ChadoConnection $chado **/
    $chado = \Drupal::service('tripal_chado.database');
    $schema = $chado->schema();
    if ($schema->tableExists($linker)) {
      return TRUE;
    }
    return FALSE;
  }
```

If the `isCompatible()` function returns FALSE then the `ChadoFieldItemBase::storageSettingsForm()` can prevent the field from being added.

This solution would mean that we can remove the following:
1.  The `object_id` in the field annotations.  This was only needed in the current implementation to tell the `ChadoFieldItemBase::storageSettingsForm()` that this was a field that handled a linker.
2.  We can remove  the `ChadoFieldItemBase::getBaseTables()` function. This is no longer needed.
3.  We can remove  the `ChadoFieldItemBase::getLinkerTables()` function. This is no longer needed.

I prefer the fields being in control of what content types they allow rather than having the `ChadoFieldItemBase::storageSettingsForm()` try to figure it out.  It makes writing a field a bit more complicated because the developer has to write the `isCompatible()` function but it also makes it easier on us because we don't have to make the settings form know how to handle every possible weird field we want to make.

@laceysanderson and @dsenalik... thoughts?

**Comment by dsenalik** [comment]
1. Wow, thanks for your late night work on this, this looks like a wonderful solution!
2. I was thinking we should merge pr #1795 first if possible, so we can start with things already working on Drupal 10.2 - **however** Lacey's `base_table_dependant` changes may no longer be needed, and all the resulting changes to practically ever field 😲 - so if we make this change first I am okay with that. 
3. Your code for the linker `isCompatible` function makes some assumptions about order in the table name, and we need to handle direct links too, so this will need to be more robust, but we could write a helper function for the parent class and maybe make it almost a one-liner in the fields.
4. Another benefit, the ajax will be gone!
5. We might still need the getLinkerTables() function, because I always wanted to support the Tripal 3 case where there were two candidate linker tables in some cases, one from an add-on module. Or maybe a site admin wants to use a different linker table for a particular content type for some reason.

**Comment by dsenalik** [comment]
Trying to build a Drupal 10.2 docker fails with this error:
`#27 133.2  [error]  TypeError: array_key_exists(): Argument #2 ($array) must be of type array, Drupal\tripal\Entity\TripalEntityType given in array_key_exists() (line 121 of /var/www/drupal/web/modules/contrib/tripal/tripal/src/Services/TripalEntityTypeCollection.php) #0 /var/www/drupal/web/modules/contrib/tripal/tripal/src/Services/TripalEntityTypeCollection.php(121): array_key_exists('settings', Object(Drupal\tripal\Entity\TripalEntityType))`

I did this
```
diff --git a/tripal/src/Services/TripalEntityTypeCollection.php b/tripal/src/Services/TripalEntityTypeCollection.php
index d0471c338..fe2211fea 100644
--- a/tripal/src/Services/TripalEntityTypeCollection.php
+++ b/tripal/src/Services/TripalEntityTypeCollection.php
@@ -118,8 +118,8 @@ class TripalEntityTypeCollection implements ContainerInjectionInterface  {
 
             // Now add any third party settings.
             $settings = [];
-            if (array_key_exists('settings', $content_type)) {
-              $settings = $content_type['settings'];
+            if (property_exists($content_type, 'settings')) {
+              $settings = $content_type->get('settings');
             }
             if (!empty($settings)){
               foreach ($settings as $key => $value) {
```

**Comment by dsenalik** [comment]
This is how the field selector works in Drupal 10.2 on this branch, I call it a nice improvement. No longer is every field shown in one screen.
You first select "Chado Fields", and then all of the chado fields appear down below, and you can select your desired field.
![2024-03-fieldselector](https://github.com/tripal/tripal/assets/8419404/1c463817-be2f-45cd-ada1-33224038eba1)

I couldn't proceed because the linker method field is disabled, gonna check on that...

**Comment by spficklin** [comment]
Yeah.  In Drupal  10.2 they added a category plugin for fields which allowed me to group our fields on that page you took a screen shot of. It's much nicer.  I had to add a category to the annotations of each field and add some new YAML files

**Comment by spficklin** [comment]
Oh also,  the branch is not fully implemented which is why you couldn't proceed.  I had to comment out some code in the form but I didn't want to keep going if there wasn't agreement with this solution.  Regarding adding a linker select if there are multiple linker tables to choose from.... I think we can let the fields take care of that too rather than try to make the base class do it.  Fields can override the settings form.

**Comment by dsenalik** [comment]
Hopefully we can discuss Monday at cofest - If we merge in some my Drupal 10.2 fixes in #1795 so that it works on that version too, this will be a huge overall improvement for the "manual field adding process".

**Comment by laceysanderson** [comment]
Thanks for your work on this @spficklin!

1. Using 3rd party settings to allow field YAML to set the base table is a great solution!
2. As mentioned in slack, we do need to keep the base table drop down since not all fields are added via the field YAML as you can also create content types and add fields through the UI.
3. I love passing off the decision of compatibility to the fields! This will make things much simpler as an extension module developer as in Tripal3 quite a few of our modules had nested ifs in the field type/instance hooks to decide when a field should be added.

I'm not sure where the discussion in slack ended up regarding the `object_table` annotation although I personally agree with @dsenalik that we should keep it as even if it could be determined, this simplifies the logic in the field and reduces the chance it makes the wrong choice 🙃 

I'm also not sure of what parts from #1795 need to be brought into this branch and how this effects #1790 which needs some updates related to Drupal 10.2 as well.

Heading off to cofest now though so I'll update here after based on what I find out in discussion.

**Comment by dsenalik** [comment]
To-do: When adding a new content type, we will need to provide a way to set the third party setting for base table

**Comment by laceysanderson** [comment]
After discussion in cofest we decided to 
- keep the object_table as a static variable and keep the linker dropdown
- keep the base table drop down but disable it if the third party setting is set. That way if the content type is created through the UI and the 3rd party setting is not set then you can still fall back to selecting the base table.
- create an issue to address that we don't have our own annotation classes for fields and thus cannot add things to the annotation with defaults and documentation.
- @dsenalik is going to move his subform related fixes from #1795 into this branch over the next week and then stephen will take it from there.
- The CVterms PR #1790 has been marked as draft and will be revisited after this work makes it into a PR.

Note: We currently have object_table in the annotation and it does "just work" but we will remove it from the annotation in favour of the static variables for now until the new issue I'm going to create is resolved.
